x-pack/plugin/search-business-rules/src/main/java/org/elasticsearch/xpack/searchbusinessrules/PinnedQueryBuilder.java

```diff
@@ -0,0 +1,218 @@
+        }
+        this.organicQuery = organicQuery;
+        return this;
+    }
+
+    /**
+     * @return the value set by {@link #organicQuery(QueryBuilder)}
+     */
+    public QueryBuilder organicQuery() {
+        return this.organicQuery;
+    }
+
+    /**
+     * Returns the pinned ids for the query.
+     */
+    public Set<String> ids() {
+        return this.ids;
+    }
+
+    @Override
+    protected void doXContent(XContentBuilder builder, Params params) throws IOException {
+        builder.startObject(NAME);
+        builder.startArray(IDS_FIELD.getPreferredName());
+        for (String value : ids) {
+            builder.value(value);
+        }
+        builder.endArray();
+        if (organicQuery != null) {
+            builder.field(ORGANIC_QUERY_FIELD.getPreferredName());
+            organicQuery.toXContent(builder, params);
+        }
+        printBoostAndQueryName(builder);
+        builder.endObject();
+    }
+
+    private static ObjectParser<PinnedQueryBuilder, Void> PARSER = new ObjectParser<>(NAME, () -> new PinnedQueryBuilder());
+
+    static {
+        PARSER.declareStringArray(fromList(String.class, PinnedQueryBuilder::addIds), PinnedQueryBuilder.IDS_FIELD);
+        PARSER.declareObject(PinnedQueryBuilder::organicQuery, (XContentParser p, Void c) -> {
+            return parseInnerQueryBuilder(p);
+        }, ORGANIC_QUERY_FIELD);
+        declareStandardFields(PARSER);
+    }
+
+    public static PinnedQueryBuilder fromXContent(XContentParser parser) {
+        try {
+            return PARSER.apply(parser, null);
+        } catch (IllegalArgumentException e) {
+            throw new ParsingException(parser.getTokenLocation(), e.getMessage(), e);
+        }
+    }
+
+    @Override
+    public String getWriteableName() {
+        return NAME;
+    }
+
+    @Override
+    protected QueryBuilder doRewrite(QueryRewriteContext queryShardContext) throws IOException {
+        if (ids.size() == 0 && organicQuery == null) {
+            return new MatchNoneQueryBuilder();
+        }
+        return this;
+    }
+
+    @Override
+    protected Query doToQuery(QueryShardContext context) throws IOException {
+        MappedFieldType idField = context.fieldMapper(IdFieldMapper.NAME);
+        if (idField == null) {
+            return new MatchNoDocsQuery("No mappings");
+        }
+        if (this.ids.isEmpty()) {
+            if (organicQuery == null) {
+                return Queries.newMatchNoDocsQuery("Missing ids and organic query in \"" + this.getName() + "\" query.");
+            } else {
+                return new CappedScoreQuery(organicQuery.toQuery(context), MAX_ORGANIC_SCORE);
+            }
+
+        } else {
+            BooleanQuery.Builder pinnedQueries = new BooleanQuery.Builder();
+
+            // Ensure each pin order using a Boost query with the relevant boost factor
+            int minPin = 0xfe << 23;
```

Comment:
I think it'd be easier to read `NumericUtils.floatToSortableInt(MAX_ORGANIC_SCORE)+1` and then `pinScore=NumericUtils.sortableIntToFloat(boostNum)` below.